### PR TITLE
[Maintenance] Introduce Flysystem and use it in place of Gaufrette by default

### PR DESCRIPTION
**Pull request by coldic3**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12 |
| Bug fix?        | no                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | yes |
| Related tickets | https://github.com/Sylius/Sylius/issues/13233                      |
| License         | MIT                                                          |

Using Gaufrette becomes deprecated in favor of Flysystem.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php [hunk 1]
```diff
@@ -69,6 +72,14 @@ public function load(array $configs, ContainerBuilder $container): void
                 $container->getDefinition('sylius.order_processing.order_prices_recalculator'),
             );
         }
+
+        $container->setAlias(
+            FilesystemAdapterInterface::class,
+            match($config['filesystem']['adapter']) {
+                'default', 'flysystem' => FlysystemFilesystemAdapter::class,
+                'gaufrette' => 'Sylius\Component\Core\Filesystem\Adapter\GaufretteFilesystemAdapter',
```
Comment (coldic3) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php:80]: Because `/** @psalm-suppress DeprecatedClass */` doesn't work for the `::class` constant :/

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -64,15 +64,22 @@
...
+        </service>
+        <service id="Sylius\Component\Core\Filesystem\Adapter\GaufretteFilesystemAdapter">
             <argument type="service">
                 <service class="Gaufrette\Filesystem">
                     <argument>%sylius.uploader.filesystem%</argument>
                     <factory service="knp_gaufrette.filesystem_map" method="get" />
                 </service>
             </argument>
-            <argument type="service" id="Sylius\Component\Core\Generator\ImagePathGeneratorInterface" />
         </service>
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:82]: As far as I remember we wanted with @lchrusciel to keep the original shape of `sylius.image_uploader` service (Idk if that information has been missed when I hand over this task).

@lchrusciel take a look and give your feedback on what you think about it. If you decided the original shape must be preserved I'll call Kevin and tell him my initial idea how to solve it.

Comment (coldic3) [reply src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:82]: The original shape is preserved but deprecated :P I just find it more suitable to use adapters instead of `AbstractImageUploader`. If you see the cons of this implementation, I'll open another PR 💃

Comment (jakubtobiasz) [reply src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:82]: @coldic3 As far as I remember Łukasz asked me to do it this way but let's wait for his opinion 😄

File: config/packages/flysystem.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+flysystem:
+    storages:
+        sylius.storage:
+            adapter: 'local'
+            options:
+                directory: '%sylius_core.images_dir%'
```
Comment (lchrusciel) [inline config/packages/flysystem.yaml]:
```suggestion
```
as it is already defined in `src/Sylius/Bundle/CoreBundle/Resources/config/app/config.yml`

File: src/Sylius/Component/Core/Filesystem/Adapter/GaufretteFilesystemAdapter.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Sylius\Component\Core\Filesystem\Adapter;
+
+use Gaufrette\FilesystemInterface;
+use Sylius\Component\Core\Filesystem\Exception\FileNotFoundException;
+
+/**
+ * @deprecated since version 1.12, to be removed in 2.0. Use {@link FilesystemInterface} instead.
+ */
+class GaufretteFilesystemAdapter implements FilesystemAdapterInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Filesystem/Adapter/GaufretteFilesystemAdapter.php]:
```suggestion
final class GaufretteFilesystemAdapter implements FilesystemAdapterInterface
```

File: src/Sylius/Component/Core/spec/Filesystem/Adapter/FlysystemFilesystemAdapterSpec.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace spec\Sylius\Component\Core\Filesystem\Adapter;
+
+use League\Flysystem\FilesystemOperator;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Filesystem\Adapter\FilesystemAdapterInterface;
+use Sylius\Component\Core\Filesystem\Exception\FileNotFoundException;
+
+class FlysystemFilesystemAdapterSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/Filesystem/Adapter/FlysystemFilesystemAdapterSpec.php]:
```suggestion
final class FlysystemFilesystemAdapterSpec extends ObjectBehavior
```

File: src/Sylius/Component/Core/spec/Filesystem/Adapter/FlysystemFilesystemAdapterSpec.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
+<?php
+
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/Filesystem/Adapter/FlysystemFilesystemAdapterSpec.php:2]:
```suggestion

/*
 * This file is part of the Sylius package.
 *
 * (c) Paweł Jędrzejewski
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

declare(strict_types=1);

```

File: src/Sylius/Component/Core/spec/Filesystem/Adapter/GaufretteFilesystemAdapterSpec.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+declare(strict_types=1);
+
+namespace spec\Sylius\Component\Core\Filesystem\Adapter;
+
+use Gaufrette\FilesystemInterface;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Core\Filesystem\Adapter\FilesystemAdapterInterface;
+use Sylius\Component\Core\Filesystem\Exception\FileNotFoundException;
+
+class GaufretteFilesystemAdapterSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/Filesystem/Adapter/GaufretteFilesystemAdapterSpec.php]:
```suggestion
final class GaufretteFilesystemAdapterSpec extends ObjectBehavior
```

File: src/Sylius/Component/Core/Filesystem/Adapter/FlysystemFilesystemAdapter.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Core\Filesystem\Adapter;
+
+use League\Flysystem\FilesystemOperator;
+use Sylius\Component\Core\Filesystem\Exception\FileNotFoundException;
+
+class FlysystemFilesystemAdapter implements FilesystemAdapterInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Filesystem/Adapter/FlysystemFilesystemAdapter.php]:
```suggestion
final class FlysystemFilesystemAdapter implements FilesystemAdapterInterface
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml (hunk 2 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:82]: I was blind and stubborn with forcing different implementation. Current one is nice. @jakubtobiasz your first idea was good

File: CONFLICTS.md [hunk 9]
```diff
@@ -29,7 +29,7 @@ references related issues.
    This version contains service with a wrong constructor arguments:
    `Invalid definition for service ".container.private.profiler": argument 4 of "JMS\SerializerBundle\Debug\DataCollector::__construct()" accepts "JMS\SerializerBundle\Debug\TraceableDriver", "JMS\SerializerBundle\Debug\TraceableMetadataFactory" passed.`
 
-   References: https://github.com/schmittjoh/JMSSerializerBundle/issues/844
```
Comment (NoResponseMate) [inline CONFLICTS.md:32]: Unrelated

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php (hunk 1 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php:80]:
```suggestion
                'gaufrette' => GaufretteFilesystemAdapter::class,
```

File: src/Sylius/Bundle/CoreBundle/Tests/TestKernel.php [hunk 10]
```diff
@@ -103,6 +104,7 @@ public function registerBundles(): array
             new JMSSerializerBundle(),
             new FOSRestBundle(),
             new KnpGaufretteBundle(),
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/TestKernel.php:106]: shouldn't we remove it?

File: src/Sylius/Bundle/CoreBundle/composer.json [hunk 11]
```diff
@@ -33,6 +33,7 @@
         "jms/serializer-bundle": "^3.5",
         "knplabs/gaufrette": "^0.8",
         "knplabs/knp-gaufrette-bundle": "^0.7",
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/composer.json:35]: shouldn't we remove it? But we should maintain integration with gaufrette if lib is present

File: src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php (hunk 1 above)
Comment (coldic3) [reply src/Sylius/Bundle/CoreBundle/DependencyInjection/SyliusCoreExtension.php:80]: https://github.com/Sylius/Sylius/pull/14298#discussion_r972018433 ;d
